docs/contributing/unit-test-guideline.md

```diff
@@ -0,0 +1,193 @@
+# Unit Test Guideline
+
+## About
+
+When we create the unit test for each function or package, we should consider test coverage not only code coverage.
+This guideline will help you to create good unit tests.
+
+## Unit Test
+
+Before considering the unit test, we should define what is the unit.
+Unit is a set of procedures or functions, in a procedural or functional language.
+Also, it is a class and its needed classes, in an object or object-oriented language.
+Unit testing validates the basic units of the program in isolation.
+
+## Guideline
+
+### Strategy
+
+Before creating the unit tests, we must know the essentials of strategy.
+
+1. Know what you are testing
+	- Purpose/Consumers of target function or classes or etc / designed by contract  defensive program.
+1. Test behaviors and results, not implementation
+	- ensures that tests only fail when there is an actual effect and not due to internal changes.
+1. Test one thing at a time
+	- Each test should have a clear, concise, singular objective.
+1. Make tests readable and understandable
+	- Hold test code to a similar standard as production code.
+1. Make tests deterministic
+	- A test should pass all the time or fail all the time until fixed.
+1. Make tests independent and self-sufficient
+	- Setup, execution, and verification steps in a given test should not depend on running other tests before it. To keep unit tests simple, fast running, and easy to debug, it may be necessary to isolate the class under test.
+1. Repeat yourself when necessary
+	- It is okay to violate the 'do not repeat yourself' principle if it makes tests simpler and easier to read.
+1. Measure code coverage but focus on test coverage
+	- Do not simply to archive code coverage.
+
+### Test case
+
+#### Basic
+
+It is not perfect, but we should try to cover all codes at first.
+To cover all code, the basic way is the creating test cases to complete all branches in the target unit.
+
+Let's see the below function.
+```go
+func calcSum(val ...int32) (sum int32) {
+	if len(val) == 0 {
+		return
+	} else {
+		for _, v := range val {
+			sum += v
+		}
+		return
+	}
+}
```

Comment:
Thank you for your example.
First, I understood your opinion. And I agree with you.
But the checked example code, we can also represent like this more simpler. 

```go
func calcSum(val ...int32) (sum int32) {
	for _, v := range val {
		sum += v
	}
	return
}
```
In golang, Even if we pass nil or empty as a variadic argument, it is converted to an empty slice.
Therefore how about changing the `calcSum ` method to return an error like this?

```go
func calcSum(val ...int32) (sum int32, err error) {
	if len(val) == 0 {
		return 0, errors.New("empty")
	}
	for _, v := range val {
		sum += v
	}
	return sum, nil
}
```

Because it is necessary to check for errors and the writing style is unified, I think the following sentences will be conveyed more clearly.

59-64
> When the above function is given, we should create 2 test for achieving 100% code coverage at least.
> - When val is not given
>   - In other words, the default value is given
> - When val(int32) is given
>  - `var val int32 = 1`

If I made a mistake, I am sorry :bow:

Reply:
@hlts2 I am not sure if returning error in this example is necessary or not. I think it depends on the program design.

Reply:
@kevindiu Thank you. I got it 👍 

> it depends on the program design.

I think so too. Is there any other example? 🤔

Reply:
Thx ,but why did you think so.
I just thought this is just the example code for explaining input values of base strategy.
I'd like to remove any  codes which is not related this explanation.
could you please read this draft document over all again?
and if you suggest about example code, please keep in your mind it is needed only for describe basic input patterns.

Reply:
> I just thought this is just the example code for explaining input values of base strategy.

I got it. thank you.